engine/src/main/scala/cromwell/engine/workflow/tokens/TokenQueue.scala

```diff
@@ -31,23 +34,42 @@ final case class TokenQueue(queue: Queue[ActorRef], private [tokens] val pool: T
     *
     * @return the new token queue
     */
-  def enqueue(actor: ActorRef): TokenQueue = {
-    copy(queue = queue.enqueue(actor))
+  def enqueue(placeholder: TokenQueuePlaceholder): TokenQueue = {
+    copy(queue = queue.enqueue(placeholder))
   }
 
   /**
-    * Returns a DequeuedActor if there's an element available, None otherwise.
+    * Returns a dequeue'd actor if one exists and there's a token available for it
+    * Returns an updated token queue based on this request (successful queuees get removed, hogs get sent to the back)
     */
-  def dequeueOption: Option[DequeuedActor] = {
-    for {
-      actorAndNewQueue <- queue.dequeueOption
-      (actor, newQueue) = actorAndNewQueue
-      lease <- pool.tryAcquire()
-    } yield DequeuedActor(LeasedActor(actor, lease), copy(queue = newQueue))
+  def dequeue: DequeueResult = {
+    queue.dequeueOption match {
+      case Some(actorAndNewQueue) =>
+        val (placeholder, newQueue) = actorAndNewQueue
+        val poolAcquisitionResult = pool.tryAcquire(placeholder.hogGroup)
+        unpackLeaseOptionAndDetermineNewQueue(poolAcquisitionResult, placeholder, queue, newQueue) match {
+          case UnpackedLeaseAndUpdatedQueue(Some(lease), nq) => DequeueResult(Option(LeasedActor(placeholder.actor, lease)), nq)
+          case UnpackedLeaseAndUpdatedQueue(None, nq) if !(this.queue eq nq.queue) => nq.dequeue
+          case UnpackedLeaseAndUpdatedQueue(None, _) => DequeueResult(None, this)
+        }
+      case None =>
+        DequeueResult(None, this)
+    }
+  }
+
+  private def unpackLeaseOptionAndDetermineNewQueue(result: UnhoggableTokenPoolResult,
```

Comment:
This is only called from the method above. If it were a function local to that method you could kill a lot of the noisy parameter passing.